sklearn/kernel_approximation.py

```diff
@@ -22,7 +24,158 @@
+        so we only care about the dimension of samples in X.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            Training data, where n_samples in the number of samples
+            and n_features is the number of features.
+
+        Returns
+        -------
+        self : object
+            Returns the transformer.
+        """
+
+        X = check_array(X, accept_sparse="csc")
+        random_state = check_random_state(self.random_state)
+
+        if sp.issparse(X) and self.coef0 != 0:
+            X = sp.hstack([X, np.sqrt(self.coef0) * np.ones((X.shape[0], 1))],
+                          format="csc")
+
+        elif not sp.issparse(X) and self.coef0 != 0:
+            X = np.hstack([X, np.sqrt(self.coef0)*np.ones((X.shape[0], 1))])
+
+        n_features = X.shape[1]
+
+        self.indexHash_ = random_state.randint(0, high=self.n_components,
+                                               size=(self.degree, n_features))
+
+        self.bitHash_ = random_state.choice(a=[-1., 1.],
+                                            size=(self.degree, n_features)
+                                            ).astype(np.float32)
+        return self
+
+    def transform(self, X):
+        """Generate the feature map approximation for X.
+
+        Parameters
+        ----------
+        X : {array-like}, shape (n_samples, n_features)
+            New data, where n_samples in the number of samples
+            and n_features is the number of features.
+
+        Returns
+        -------
+        X_new : array-like, shape (n_samples, n_components)
+        """
+
+        check_is_fitted(self)
+        X = check_array(X, accept_sparse="csc")
+
+        if sp.issparse(X) and self.coef0 != 0:
+            X = sp.hstack([X, np.sqrt(self.coef0)*np.ones((X.shape[0], 1))],
+                          format="csc")
+
+        elif not sp.issparse(X) and self.coef0 != 0:
+            X = np.hstack([X, np.sqrt(self.coef0)*np.ones((X.shape[0], 1))])
+
+        if X.shape[1] != self.indexHash_.shape[1]:
+            raise ValueError("Number of features of test samples does not"
+                             " match that of training samples.")
+
+        Ps = np.zeros((X.shape[0], self.degree, self.n_components))
+
+        if sp.issparse(X):
+            for j in range(X.shape[1]):
+                for d in range(self.degree):
+                    iHashIndex = self.indexHash_[d, j]
+                    iHashBit = self.bitHash_[d, j]
+                    Ps[:, d, iHashIndex] += \
+                        (iHashBit * X[:, j]).toarray().ravel()
+
+        else:
+            for j in range(X.shape[1]):
+                for d in range(self.degree):
+                    iHashIndex = self.indexHash_[d, j]
+                    iHashBit = self.bitHash_[d, j]
+                    Ps[:, d, iHashIndex] += iHashBit * X[:, j]
+
+        Ps = fftpack.fft(Ps, axis=2, overwrite_x=True)
```

Comment:
I would avoid reusing the same variable name `Ps`, for the sake of clarity.

Reply:
- [ ] Give intermediate variables more meaningful names.